autonat/autonat-v2.md

```diff
@@ -0,0 +1,255 @@
+ResponseStatus](#requirements-for-responsestatus).
+
+The initiator MUST check that the nonce received in the `DialAttempt` is the
+same as the nonce the it sent in the `DialRequest`. If the nonce is different,
+it MUST discard this response.
+
+The peer MUST close the stream after sending the response. The initiator MUST
+close the stream after receiving the response.
+
+
+### Requirements for DialStatus
+
+On receiving a `DialRequest` the peer goes through the list of addresses in the
+request to select the first address that it is capable of dialing. For every
+address that the peer checks, it assigns a `DialStatus` according to the
+following requirements. 
+
+For addresses that the peer decides to not dial:
+
+`E_ADDRESS_UNKNOWN`: The peer didn't understand the address. 
+
+`E_TRANSPORT_NOT_SUPPORTED`: The peer understood the address, but has no
+transport capable of dialing the requested address. 
+
+`E_DIAL_REFUSED`: The peer didn't dial the address because of address
+based restrictions like address based rate limit, the address being a private IP
+address, or a relay address.
+
+For the address that the peer decided to dial:
+
+`E_DIAL_ERROR`: The peer was unable to connect to the address
+
+`E_CONN_UPGRADE_FAILED`: The peer was able to connect to the other address, but
+failed to complete the connection upgrade step. 
+
+`E_ATTEMPT_ERROR`: The peer was able to establish an upgraded connection but
+some error occured when sending the nonce on the `/libp2p/autonat/2/attempt`
+stream.
+
+`OK`: The peer successfully dialed the selected address.
+
+The peer MUST NOT send dial statuses for addresses after the one it selected to
+dial.
+
+
+### Requirements for ResponseStatus
+
+The `ResponseStatus` sent by the peer in the `DialResponse` message MUST be set
+according to the following requirements
+
+`OK`: the peer completed the request successfully. A request is considered
+completed successfully when the peer either completes a dial(successfully or
+unsuccessfully) or rejects all addresses in the request as undialable. 
+
+`E_BAD_REQUEST`: the peer was unable to decode the request.
+
+`E_REQUEST_REFUSED`: the peer didn't attempt to serve the request because of
+rate limiting, resource limit reached or blacklisting.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand. 
+
+### Amplification Attack Prevention
+
+![Interaction](autonat-v2-amplification-attack-prevention.svg)
+
+
+When a client asks a server to dial an address that is not the client's observed
+IP address, the server asks the client to send him some non trivial amount of
+bytes as a cost to dial a different IP address. To make amplification attacks
+unattractive, the number of bytes is decided such that it's sufficiently larger
+than a new connection handshake cost.
```

Comment:
@sukunrt @marten-seemann do you have a recommendation on the number of bytes to request in mind?

Reply:
We're talking about an amplification attack, so number of bytes sent is the currency we should be thinking in.

If you want to introduce a PoW, you'd have to think about the exchange rate between CPU cycles and bandwidth, which is not at all obvious what a safe value would be.

Reply:
The real cost here is getting the Server to do a handshake with a target. see @mxinden's [comment](https://github.com/libp2p/specs/issues/536#issuecomment-1505313751) for details. 

How should we measure the compute cost of a TLS or noise handshake?

Reply:
The above reasoning makes a lot of sense. Thanks @sukunrt and @marten-seemann.

I would expect future readers to ask the same question as @thomaseizinger did above. Can you include your reasoning in the document? I have found an `FAQ` section at the end of a specification useful, though I don't feel strongly about that structuring.

See e.g. https://github.com/libp2p/specs/blob/master/webrtc/webrtc.md#faq.

> do you have a recommendation on the number of bytes to request in mind?

I still think this is helpful for implementers.

Reply:
>Can you include your reasoning in the document? 
Will do.

>do you have a recommendation on the number of bytes to request in mind?

I'll add this to the doc after calculating/measuring.
Off the top of my head, I think most handshakes finish within 10k bytes exchanged so something like 30KB-100KB should be enough.

Reply:
added the recommendation.

Reply:
>To make amplification attacks
unattractive, servers SHOULD ask for 30k to 100k bytes. Since most handshakes
cost less than 10k bytes in bandwidth, 30kB is sufficient to make attacks
unattractive.